### PR TITLE
Update karmor recommend command to include namespace

### PR DESCRIPTION
**Pull request by vishnusomank**
- Added use-namespace flag to input namespace

- Change policy name from default to container specific [ [NAMESPACE-]CONTAINERIMGNAME-POLICY-NAME ]

Addresses: https://github.com/kubearmor/kubearmor-client/issues/112#issuecomment-1216476912

Signed-off-by: vishnusomank <vishnu@accuknox.com>

### CODE REVIEW

File: recommend/recommend.go [hunk 1]
```diff
@@ -17,11 +17,12 @@ import (
...
-	Uselabels  []string
-	Tags       []string
-	Outdir     string
-	Reportfile string
+	Images       []string
+	Uselabels    []string
+	Tags         []string
+	Usenamespace string
+	Outdir       string
+	Reportfile   string
```
Comment (wazir-ahmed) [inline recommend/recommend.go:25]: Naming should be always in UpperCamelCase. `UseLabels`, `UserNamespace`, `OutDir` and `ReportFile`.

File: cmd/recommend.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ func init() {
 
 	recommendCmd.Flags().StringSliceVar(&recommendOptions.Images, "image", []string{}, "Container image list (comma separated)")
 	recommendCmd.Flags().StringSliceVar(&recommendOptions.Uselabels, "use-labels", []string{}, "User defined labels for policy (comma separated)")
+	recommendCmd.Flags().StringVarP(&recommendOptions.Usenamespace, "use-namespace", "", "", "User defined namespace value for policies")
```
Comment (wazir-ahmed) [inline cmd/recommend.go:37]: We can have shorthand options here.
- `-i` for `--image`
- `-l` for `--use-labels`
- `-n` for `--use-namespace`

File: recommend/policy.go [hunk 3]
```diff
@@ -77,9 +77,14 @@ func (img *ImageInfo) createPolicy(ms MatchSpec) (types.KubeArmorPolicy, error)
 
 	//	policy.Metadata["containername"] = img.RepoTags[0]
 	if ms.Name != "" {
-		policy.Metadata["name"] = ms.Name
+		policy.Metadata["name"] = strings.TrimPrefix(strings.TrimSpace(options.Usenamespace+"-"+mkPathFromTag(img.RepoTags[0])+"-"+ms.Name), "-")
```
Comment (wazir-ahmed) [inline recommend/policy.go:80]: Declutter this line.

```suggestion
		policy.Metadata["name"] = strings.TrimPrefix(strings.TrimSpace(options.Usenamespace + "-" + mkPathFromTag(img.RepoTags[0]) + "-" + ms.Name), "-")
```
---
Also why we need to use `strings.TrimSpace()` here?

File: recommend/policy.go [hunk 4]
```diff
@@ -77,9 +77,14 @@ func (img *ImageInfo) createPolicy(ms MatchSpec) (types.KubeArmorPolicy, error)
...
 	} else {
-		policy.Metadata["name"] = "ksp-" + mkPathFromTag(img.RepoTags[0])
+		policy.Metadata["name"] = "ksp-" + strings.TrimPrefix(strings.TrimSpace(options.Usenamespace+"-"+mkPathFromTag(img.RepoTags[0])), "-")
+	}
+
+	// Condition to set namespace, if user defined namespace value is available
+	if len(options.Usenamespace) > 0 {
```
Comment (wazir-ahmed) [inline recommend/policy.go:86]:
```suggestion
	if options.Usenamespace != "" {
```

File: recommend/policy.go [hunk 5]
```diff
@@ -77,9 +77,14 @@ func (img *ImageInfo) createPolicy(ms MatchSpec) (types.KubeArmorPolicy, error)
 
 	//	policy.Metadata["containername"] = img.RepoTags[0]
 	if ms.Name != "" {
-		policy.Metadata["name"] = ms.Name
+		policy.Metadata["name"] = strings.TrimPrefix(strings.TrimSpace(options.Usenamespace+"-"+mkPathFromTag(img.RepoTags[0])+"-"+ms.Name), "-")
 	} else {
-		policy.Metadata["name"] = "ksp-" + mkPathFromTag(img.RepoTags[0])
+		policy.Metadata["name"] = "ksp-" + strings.TrimPrefix(strings.TrimSpace(options.Usenamespace+"-"+mkPathFromTag(img.RepoTags[0])), "-")
```
Comment (wazir-ahmed) [inline recommend/policy.go:82]: Declutter this line too. Use `fmt.Sprintf`

**Review by wazir-ahmed — CHANGES_REQUESTED** [review]
Please address the minor comments.

File: cmd/recommend.go (hunk 2 above)
Comment (nyrahul) [reply cmd/recommend.go]: Hey @wazir-ahmed , use-labels and use-namespace should not use the same shorthand imo ... We will need those shorthands subsequently for k8s based recommendations. CC: @vishnusomank

File: cmd/recommend.go [hunk 6]
```diff
@@ -32,9 +32,10 @@ var recommendCmd = &cobra.Command{
 func init() {
 	rootCmd.AddCommand(recommendCmd)
 
-	recommendCmd.Flags().StringSliceVar(&recommendOptions.Images, "image", []string{}, "Container image list (comma separated)")
-	recommendCmd.Flags().StringSliceVar(&recommendOptions.Uselabels, "use-labels", []string{}, "User defined labels for policy (comma separated)")
-	recommendCmd.Flags().StringVarP(&recommendOptions.Outdir, "outdir", "o", "out", "output folder to write policies")
-	recommendCmd.Flags().StringVarP(&recommendOptions.Reportfile, "report", "r", "report.txt", "report file")
+	recommendCmd.Flags().StringSliceVarP(&recommendOptions.Images, "image", "i", []string{}, "Container image list (comma separated)")
+	recommendCmd.Flags().StringSliceVarP(&recommendOptions.UseLabels, "use-labels", "l", []string{}, "User defined labels for policy (comma separated)")
+	recommendCmd.Flags().StringVarP(&recommendOptions.UseNamespace, "use-namespace", "n", "", "User defined namespace value for policies")
```
Comment (nyrahul) [inline cmd/recommend.go:37]: Sorry for this, but please remove use of the shorthands here.

These shorthands would be required for k8s based recommendations.

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: cmd/recommend.go (hunk 6 above)
Comment (vishnusomank) [reply cmd/recommend.go]: done @nyrahul

File: cmd/recommend.go (hunk 2 above)
Comment (wazir-ahmed) [reply cmd/recommend.go]: @nyrahul What do you think about this.. 
- We will remove `use-labels` and `use-namespace` options. And add `labels` and `namespace` options.
- If user invokes the `recommend` command with `--image`, then we consider it as container-image based recommendation.
```
karmor recommend -i ubuntu:20.04 -l app=ubuntu -n test
```
If the user invokes the command without `--image` option, then we consider it as k8s based recommendation.  
```
karmor recommend -l app=feeder -n test
```

Comment (nyrahul) [reply cmd/recommend.go]: That's a good proposition. @vishnusomank, WDYT?

Comment (vishnusomank) [reply cmd/recommend.go]: Sounds good. This will also make the user experience better. With some minor adjustments, we can achieve this too

**Review by wazir-ahmed — APPROVED** [review]

File: cmd/recommend.go (hunk 2 above)
Comment (wazir-ahmed) [reply cmd/recommend.go]: @nyrahul 
I will add the above mentioned changes as part of k8s manifest changes. The PR is in good shape. We can merge this.

**Review by nyrahul — APPROVED** [review]

**Pull request merged**